backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/ShipmentLineModularContractHandler.java

```diff
@@ -0,0 +1,190 @@
+	public @NonNull Optional<LogEntryCreateRequest> createLogEntryCreateRequest(final @NonNull I_M_InOutLine inOutLineRecord, final @NonNull FlatrateTermId flatrateTermId)
+	{
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(flatrateTermId);
+		if (modularContractSettings == null)
+		{
+			return Optional.empty();
+		}
+
+		final I_M_InOut inOutRecord = inoutDao.getById(InOutId.ofRepoId(inOutLineRecord.getM_InOut_ID()));
+		final I_C_Flatrate_Term flatrateTermRecord = flatrateDAO.getById(flatrateTermId);
+		final I_C_UOM uomId = uomDAO.getById(UomId.ofRepoId(inOutLineRecord.getC_UOM_ID()));
+		final Quantity quantity = Quantity.of(inOutLineRecord.getMovementQty(), uomId);
+
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
+				.stream()
+				.filter(config -> config.isMatchingClassName(ShipmentLineModularContractHandler.class.getName()))
+				.map(ModuleConfig::getModularContractType)
+				.map(ModularContractType::getId)
+				.findFirst();
+
+		return modularContractTypeId.map(contractTypeId -> LogEntryCreateRequest.builder()
+				.contractId(flatrateTermId)
+				.productId(ProductId.ofRepoId(inOutLineRecord.getM_Product_ID()))
+				.referencedRecord(TableRecordReference.of(I_M_InOutLine.Table_Name, inOutLineRecord.getM_InOutLine_ID()))
+				.collectionPointBPartnerId(BPartnerId.ofRepoId(inOutRecord.getC_BPartner_ID()))
+				.producerBPartnerId(BPartnerId.ofRepoId(inOutRecord.getC_BPartner_ID()))
+				.invoicingBPartnerId(BPartnerId.ofRepoId(flatrateTermRecord.getBill_BPartner_ID()))
+				.warehouseId(WarehouseId.ofRepoId(inOutRecord.getM_Warehouse_ID()))
+				.documentType(LogEntryDocumentType.SHIPMENT)
+				.soTrx(SOTrx.SALES)
+				.processed(false)
+				.quantity(quantity)
+				.amount(null)
+				.transactionDate(LocalDateAndOrgId.ofTimestamp(inOutRecord.getMovementDate(), OrgId.ofRepoId(inOutLineRecord.getAD_Org_ID()), orgDAO::getTimeZone))
+				.year(modularContractSettings.getYearAndCalendarId().yearId())
+				.description("Shipment completed")
+				.modularContractTypeId(contractTypeId)
+				.build());
+	}
+
+	@Override
+	public @NonNull Optional<LogEntryReverseRequest> createLogEntryReverseRequest(final @NonNull I_M_InOutLine inOutLineRecord, final @NonNull FlatrateTermId flatrateTermId)
+	{
+		return Optional.of(LogEntryReverseRequest.builder()
+								   .referencedModel(TableRecordReference.of(I_M_InOutLine.Table_Name, inOutLineRecord.getM_InOutLine_ID()))
+								   .flatrateTermId(flatrateTermId)
+								   .description("Shipment reversed")
+								   .build());
+	}
+
+	@Override
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_M_InOutLine inOutLineRecord)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(inOutLineRecord.getC_Order_ID()));
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID());
+		assert warehouseId != null;
```

Comment:
we don't use java-1.4. asserts, i don't even know that they are (always) swtiched on. 
Since order.getM_Warehouse_ID is a mandatory column, please just use `WarehouseId.ofRepoId(....)`